sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelHealthChecker.java

```diff
@@ -304,17 +303,81 @@ private String isReadHang(Timestamps timestamps, long currentTime, RntbdRequestM
+            if (readDelay >= this.timeoutTimeLimitInNanos) {
+                transitTimeoutValidationMessage = MessageFormat.format(
+                    "{0} health check failed due to transit timeout detection time limit: [rntbdContext: {1},"
+                        + "lastChannelRead: {2}, timeoutTimeLimitInNanos: {3}]",
+                    channel,
+                    rntbdContext,
+                    timestamps.lastReadTime,
+                    this.timeoutTimeLimitInNanos);
+
+                logger.warn(transitTimeoutValidationMessage);
+                return transitTimeoutValidationMessage;
+            }
+
+            // Consecutive connection timeout happens
+            // Max(TransitTimeoutDetectionThreshold * networkRequestTimeout, transitTimeoutHighFrequencyGracePeriodInNanos)
+            if (timestamps.transitTimeoutCount() >= this.timeoutHighFrequencyThreshold
+                && readDelay >= this.timeoutHighFrequencyTimeLimitInNanos) {
+                transitTimeoutValidationMessage = MessageFormat.format(
+                    "{0} health check failed due to transit timeout high frequency threshold hit: [rntbdContext: {1},"
+                        + "lastChannelRead: {2}, transitTimeoutCount: {3}, timeoutHighFrequencyThreshold: {4}, timeoutHighFrequencyTimeLimitInNanos: {5}]",
+                    channel,
+                    rntbdContext,
+                    timestamps.lastReadTime,
+                    timestamps.transitTimeoutCount,
+                    this.timeoutHighFrequencyThreshold,
+                    this.timeoutHighFrequencyTimeLimitInNanos);
+
+                logger.warn(transitTimeoutValidationMessage);
+                return transitTimeoutValidationMessage;
+            }
+
+            // timeout happens on write operations
+            // For write operation, SDK can only send request to primary replica, so the detection here will be more aggressive
+            if (timestamps.tansitTimeoutWriteCount() >= this.timeoutOnWriteThreshold
+                && readDelay >= this.timeoutOnWriteTimeLimitInNanos) {
+                transitTimeoutValidationMessage = MessageFormat.format(
+                    "{0} health check failed due to transit timeout on write threshold hit: [rntbdContext: {1},"
+                        + "lastChannelRead: {2}, transitTimeoutWriteCount: {3}, timeoutOnWriteThreshold: {4}, timeoutOnWriteTimeLimitInNanos: {5}]",
+                    channel,
+                    rntbdContext,
+                    timestamps.lastReadTime,
+                    timestamps.transitTimeoutWriteCount,
+                    this.timeoutOnWriteThreshold,
+                    this.timeoutOnWriteTimeLimitInNanos);
+
+                logger.warn(transitTimeoutValidationMessage);
+                return transitTimeoutValidationMessage;
+            }
+
+            //  case 4: TO Be implemented - Can we defer the health status based on the RntbdServiceEndpointStats?
```

Comment:
I assume what you are saying if on the ServiceEndpoint we had successful requests very recently let's throw away the failed channel rather faster because chances are good that other channels are healthy?

Reply:
Removed it from this PR as it the timeoutDetectionTimeLimit check should already cover part of this case as well.